kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -414,250 +415,150 @@ public static Tuple2<Integer, Integer> minimumRequiredVersions(Set<TableFeature>
+   * to know more about "auto-update capable" features.
    *
-   * @param protocol Table protocol
-   * @param metadata Table metadata
+   * <p>Note: this method only considers metadata-enabled features. To avoid confusion, the caller
+   * must apply and remove protocol-related table properties from the metadata before calling this
+   * method.
    */
-  public static void validateWriteSupportedTable(
-      Protocol protocol, Metadata metadata, String tablePath) {
-    int minWriterVersion = protocol.getMinWriterVersion();
-    switch (minWriterVersion) {
-      case 1:
-        break;
-      case 2:
-        // Append-only and column invariants are the writer features added in version 2
-        // Append-only is supported, but not the invariants
-        validateNoInvariants(metadata.getSchema());
-        break;
-      case 3:
-        // Check constraints are added in version 3
-        throw unsupportedWriterProtocol(tablePath, minWriterVersion);
-      case 4:
-        // CDF and generated columns are writer features added in version 4
-        throw unsupportedWriterProtocol(tablePath, minWriterVersion);
-      case 5:
-        // Column mapping is the only one writer feature added in version 5
-        throw unsupportedWriterProtocol(tablePath, minWriterVersion);
-      case 6:
-        // Identity is the only one writer feature added in version 6
-        throw unsupportedWriterProtocol(tablePath, minWriterVersion);
-      case 7:
-        for (String writerFeature : protocol.getWriterFeatures()) {
-          if (writerFeature.equals(INVARIANTS_FEATURE_NAME)) {
-            // For version 7, we allow 'invariants' to be present in the protocol's writerFeatures
-            // to unblock certain use cases, provided that no invariants are defined in the schema.
-            validateNoInvariants(metadata.getSchema());
-          } else if (!SUPPORTED_WRITER_FEATURES.contains(writerFeature)) {
-            throw unsupportedWriterFeature(tablePath, writerFeature);
-          }
-        }
-
-        // Eventually we may have a way to declare and enforce dependencies between features.
-        // By putting this check for row tracking here, it makes it easier to spot that row
-        // tracking defines such a dependency that can be implicitly checked.
-        if (isRowTrackingSupported(protocol) && !isDomainMetadataSupported(protocol)) {
-          throw DeltaErrors.rowTrackingSupportedWithDomainMetadataUnsupported();
-        }
-        break;
-      default:
-        throw unsupportedWriterProtocol(tablePath, minWriterVersion);
+  public Optional<Protocol> upgradeProtocolFromMetadataForExistingTable(
+      Metadata metadata, Protocol current) {
+
+    Protocol required =
+        new Protocol(TABLE_FEATURES_MIN_READER_VERSION, TABLE_FEATURES_MIN_WRITER_VERSION)
+            .withFeatures(extractAutomaticallyEnabledFeatures(metadata, current))
+            .normalized();
+
+    if (!required.canUpgradeTo(current)) {
```

Comment:
wouldn't we want to see if the `currProtocol` can be upgraded to the `requiredProtocol` ?